simphony/io/tests/test_h5_cuds.py

```diff
@@ -255,5 +259,85 @@ def tearDown(self):
             engine.close()
 
 
+class TestMaterialsOperations(unittest.TestCase):
+
+    def setUp(self):
+        self.temp_dir = tempfile.mkdtemp()
+        self.filename = os.path.join(self.temp_dir, 'test.cuds')
+        self.addTypeEqualityFunc(
+            Material, partial(compare_materials, testcase=self))
+
+        self.expected = Materials()
+        for i in xrange(5):
+            self.expected.add_material(
+                Material(description="Material {}".format(i),
+                         data=create_data_container()))
+
+    def tearDown(self):
+        shutil.rmtree(self.temp_dir)
+
+    def file_factory(self):
+        return H5CUDS.open(self.filename)
+
+    def cleanup(self):
+        os.remove(self.filename)
+
+    def test_add_get_materials(self):
+        handle = self.file_factory()
+        expected_materials_dict = {}
+        for material in self.expected.iter_materials():
+            expected_materials_dict[material.uid] = material
+
+        handle.add_materials(self.expected)
+
+        # Closing and reopening the file
+        handle.close()
+        handle = self.file_factory()
+
+        stored = handle.get_materials()
+
+        stored_materials = [material for material in stored.iter_materials()]
+
+        self.assertEqual(len(expected_materials_dict), len(stored_materials))
+        for material in stored_materials:
+            self.assertEqual(
+                material, expected_materials_dict[material.uid])
+
+        handle.close()
+        self.cleanup()
+
+    def test_remove_materials(self):
+        handle = self.file_factory()
+        handle.add_materials(self.expected)
+
+        handle.remove_materials()
+
+        # Closing and reopening the file
+        handle.close()
+        handle = self.file_factory()
+
+        stored = handle.get_materials()
+        stored_materials = [material for material in stored.iter_materials()]
+
+        # The file should not contain any materials
+        self.assertEqual(len(stored_materials), 0)
+
+        handle.close()
+        self.cleanup()
+
+    def test_error_when_adding_existing_materials(self):
+        handle = self.file_factory()
+        handle.add_materials(self.expected)
+
+        handle.close()
+        handle = self.file_factory()
+
+        with self.assertRaises(ValueError):
+            handle.add_materials(self.expected)
+
+        handle.close()
+        self.cleanup()
+
+
```

Comment:
>Maybe there's something in the big picture that I'm missing, so follow-up comments are very much appreciated.

Mostly, I feel that `MaterialRelations` is a different beast than the datasets. 

 For our datasets we defined abstract base classes (`ABCMesh`, `ABCParticles`, `ABCLattice`) and then have concrete implementations (e.g. `Particles`, `H5Particles`, `LammpsPaticles`...) that provide this "proxy" behavior (*"the a CUDS container is returned as a result of a ``get`` operation in a modelling engine the returned Container is a proxy to the information stored in the internal memory if the wrapper/modelling engine."*).  This allows for optimizations in each of the implementations for certain situations (e.g. a mesh had millions of points but user is only accessing a few hundred of them).

For `MaterialRelations`, we don't have such defined behavior as the class represents something much smaller (~ a few dozen materials).  So creating `H5Materials` seems unneeded.

My suggestion to remove `H5Materials` is just to simplify things in the short term.  I see your point about cluttering `H5Cuds` though.  Regardless, we shouldn't spend to much time thinking about it.  If others are happy with the current approach, lets just merge :)  

(if we do keep H5Materials, we need to rework the testing so that we don't repeat ourselves in the testing performed in `simphony/io/tests/test_h5_materials.py`` and 'simphony/cuds/tests/test_materials.py')